README.md
# Intro

The PAM Duress is a module designed to allow users to generate 'duress' passwords that when used in place of their
normal password will execute arbitrary scripts.

This functionality could be used to allow someone pressed to give a password under coercion to provide a password that
grants access but in the background runs scripts to clean up sensitive data, close connections to other networks to
limit lateral movement, and/or to send off a notification or alert (potentially one with detailed information like
location, visible wifi hot-spots, a picture from the camera, a link to a stream from the microphone, etc). You could
even spawn a process to remove the pam_duress module so the threat actor won't be able to see if the duress module was
available.

This is transparent to the person coercing the password from the user as the duress password will grant authentication
and drop to the user's shell.

Duress scripts can be generated on an individual user basis or generated globally. Users can also re-use global duress
passwords to sign their own duress scripts (rare instance where this could actually be useful from a security
perspective).

Contributions to this project are more than welcome; refer to our guidance on making contributions
[here](docs/Contributing.md).

**Maintainer's Notes:** While this project is primarily developed in Debian
flavors of linux, the maintainer's will make every attempt to spin up VM's or
docker containers to troubleshoot linux specific flavor issues. We also
welcome contributions from the community. Below are specific linux variant
installation guides available:

- [Arch Linux](docs/ArchLinux.md)

## Demo

[![Video Demo - https://www.youtube.com/watch?v=ShVYxhHUYdg](https://img.youtube.com/vi/ShVYxhHUYdg/0.jpg)](https://www.youtube.com/watch?v=ShVYxhHUYdg)

## Requirements

```bash
# Ubuntu/Debian dependencies
sudo apt-get install build-essential libpam0g-dev libssl-dev
```

## Build

```bash
make
sudo make install
make clean
# make uninstall
```

### Debug Build

```bash
# Debug build provides detailed output to syslog.\
sudo make uninstall
make clean
make debug
sudo make install
```

**NOTE**: In debug builds script output IS NOT redirected to /dev/null by default; in non-debug builds it is.

## Configuration

Configuration of the duress module is split into two different configuration directories. After installation, you'll
need to manually create both of them.

```bash
mkdir -p ~/.duress # Local duress scripts.
sudo mkdir -p /etc/duress.d  # Global Duress scripts.
```

After creating the directories you can create script and put them in these directories. To assign a password to execute
a particular script you use the duress_sign to create a combination password hash and integrity hash for the script.

```bash
$> duress_sign ~/.duress/delete_workspace.sh
Password: 
Confirm: 
Reading /home/user/.duress/delete_workspace.sh, 33...
Done
6B8B621EFB8050B83AAC734D56BF9165DC55D709CBAD530C6241E8A352587B3F
$> chmod -R 500 ~/.duress
$> ls -al ~/.duress/
drwxr-xr-x 2 user user 4096 Aug 20 15:15 .
drwxr-xr-x 8 user user 4096 Aug 20 15:11 ..
-r-x------ 1 user user   33 Aug 20 15:11 delete_workspace.sh
-r-x------ 1 user user   32 Aug 20 21:49 delete_workspace.sh.sha256
```

**NOTE:** Scripts will only execute with permission masks of 500, 540, or 550

**NOTE:** User generated duress scripts are only run when they attempt to log in AND use a duress password that one of
their scripts is signed with. If user Jill signs their scripts with the same password as a global script, when they use
it the global scripts will run, followed by Jill's duress scripts, but Bob, Jane, or Dan's scripts will not be run even
if they also re-used the same duress scripts.

## PAM Configuration

Modify /etc/pam.d/common-auth from the following defaults:

```bash
auth    [success=1 default=ignore]      pam_unix.so

auth    requisite                       pam_deny.so
```

To the below:

```bash
# Example /etc/pam.d/common-auth
auth    [success=2 default=ignore]      pam_unix.so
auth    [success=1 default=ignore]      pam_duress.so

auth    requisite                       pam_deny.so
```

### Order of Operations Normal Password

- User enters their standard username and password.
- pam_unix.o confirms them and returns PAM_SUCESS and skips 2 past pam_deny.o.

### Order of Operations Duress Password

- The pam_unix.o module first checks standard username and password, but since the duress password is not the users
  actual password it fails resulting in a default behavior of 'ignore' per the configuration.
- PAM then applies the username/password to pam_duress.so which:
  - Enumerates files in /etc/duress.d/
  - Checks for files that have matching .sha256 extensions
  - Hashes the provided password salted with the sha256 hash of the file and compares it with the one stored in the
    .sha256 extension file
  - If the hashes match, the script is executed via:
    - `export PAMUSER=[USERNAME]; /bin/sh [FILE]` for /etc/duress.d scripts (run as root).
    - `export PAMUSER=[USERNAME]; su - [USERNAME] -c "/bin/sh [FILE]"` for ~/.duress scripts (run as the user).
    - NOTE: PAMUSER is set so global duress scripts can specify the account flagging duress.
  - Process is repeated for all files in ~/.duress/ for the user attempting to log in.
  - Finally if ANY script is run, PAM_SUCCESS is return. Otherwise PAM_IGNORE is returned.
- If PAM_SUCESS is returned PAM will skip 1 and move past pam_deny.o to continue the pam module processes, eventually
  dropping to an authenticated shell. Otherwise the default 'ignore' behavior is honored moving to pam_deny.o,
  resulting in a failed authentication.

![Simple Flow Diagram](docs/img/diagram.png "Simple Flow Diagram")

## Testing

It is easy to do a quick test to ensure the duress module is working properly.

```bash
$> mkdir -p ~/.duress
$> echo '#!/bin/sh
echo "Hello World"' > ~/.duress/hello.sh
$> duress_sign ~/.duress/hello.sh
Password: # Enter a duress password that is NOT your actual password.
Confirm: 
$> chmod 500 ~/.duress/hello.sh
$> chmod 400 ~/.duress/hello.sh.sha256
$> sudo pam_test $USER
Credentials accepted.
Password: # Enter the password you signed the hello.sh script with.
Hello World # This output is from the duress script...
Account is valid. # ...and we still got a valid authentication.
Authenticated
$> sudo pam_test $USER
Credentials accepted.
Password: # Now enter your actual password.
Account is valid. # Note, Hello World doesn't print.
Authenticated
```

## Example Implementations

- [Use Pushover to Notify IT You're Under Duress](docs/examples/Pushover.md)

## Reference

- Initial starter code and test code from [beatgammit/simple-pam](https://github.com/beatgammit/simple-pam).
- Contribution guidelines built from template provided by
  [briandk/CONTRIBUTING.md](https://gist.github.com/briandk/3d2e8b3ec8daf5a27a62).



CHANGELOG.md
# Changelog

- 2.0.0
  - **BREAKING CHANGE:** Changed run order of scripts to run user scripts first and system scripts after. This enables
    the possibility of uninstalling pam-duress as a system level operation.
  - Linted with clang-tidy. No linter errors in files.
  - Changed outputs of errno to strerror(errno).
  - General spelling pass.
  - 20220404 - DOC ONLY UPDATE: Added Arch Linux Documentation
- 1.1.7
  - Removal of exit statements; replaced with NULL to prevent the PAM application exiting. Contribution by
    [Jürgen Hötzel on github](https://github.com/juergenhoetzel).
- 1.1.6
  - Removal of unnecessary intermediate shell. Contribution by
    [Jürgen Hötzel on github](https://github.com/juergenhoetzel).
  - Debug builds will not redirect output of stderr and stdout to /dev/null by default to support testing/debugging.
- 1.1.5
  - Makefile improvements by [Prateek Ganguli on github](https://github.com/pganguli). Debug build path now added.
- 1.1.4
  - Memory leak resolved by [Jürgen Hötzel on github](https://github.com/juergenhoetzel).
- 1.1.3
  - Fixed privilege escalation issue #16 [reported by wowaname](https://news.ycombinator.com/item?id=28276200) from
    Hacker News.
  - Redirected output of all scripts/binaries to /dev/null by convention.
- 1.1.2
  - Fixed issue #11; error when running scripts under /etc/duress.d resulting in account unavailable error.
  - Fixed issue #10; documentation or pushover script.
  - Added link to demo video.
  - Removed unnecessary casts for malloc calls.
  - Created dbg_log wrapper function to clean up DEBUG compile flag use.
- 1.1.1
  - OSX support and makefile improvements contributed by [cormacrelf](https://github.com/cormacrelf).
- 1.1.0
  - Fixed privilege escalation vulnerability that could allow an unprivileged user to run commands as root.
- 1.0.1
  - Fixed some potential memory leaks, linted, and adjusted documentation.
- 1.0.0
  - Initial commit of prototype tested on Debian 10.


src/pam_test.c
#include <security/pam_appl.h>

/* macOS uses openpam. https://stackoverflow.com/a/66853251 */
#ifdef   OPENPAM
#include <security/openpam.h>
#define  USE_CONV_FUNC  openpam_ttyconv
#else
#include <security/pam_misc.h>
#define  USE_CONV_FUNC  misc_conv
#endif

#include <stdlib.h>
#include <stdio.h>
#include "version.h"

const struct pam_conv conv = {
	USE_CONV_FUNC,
	NULL};

int main(int argc, char *argv[])
{
	pam_handle_t *pamh = NULL;
	int retval;
	const char *user = "nobody";

	if (argc != 2)
	{
		printf("Usage: app [username]\n");
		exit(1);
	}

	user = argv[1];

	retval = pam_start("check_user", user, &conv, &pamh);

	/* Are the credentials correct? */
	if (retval == PAM_SUCCESS)
	{
		printf("Credentials accepted.\n");
		retval = pam_authenticate(pamh, 0);
	}

	/* Can the account be used at this time? */
	if (retval == PAM_SUCCESS)
	{
		printf("Account is valid.\n");
		retval = pam_acct_mgmt(pamh, 0);
	}

	/* Did everything work? */
	if (retval == PAM_SUCCESS)
	{
		printf("Authenticated\n");
	}
	else
	{
		printf("Not Authenticated\n");
	}

	/* close PAM (end session) */
	if (pam_end(pamh, retval) != PAM_SUCCESS)
	{
		pamh = NULL;
		printf("check_user: failed to release authenticator\n");
		exit(1);
	}

	return retval == PAM_SUCCESS ? 0 : 1;
}


src/util.c
#include "util.h"

unsigned long get_file_size(FILE *fp)
{
    unsigned long ret = 0;
    rewind(fp);
    fseek(fp, 0L, SEEK_END);
    ret = ftell(fp);
    rewind(fp);
    return ret;
}

char *get_hash_filename(const char *filepath)
{
    size_t hash_filename_len = strlen(filepath) + strlen(SIGNATURE_EXTENSION);
    char *hash_filename = malloc(hash_filename_len + 1);
    if (hash_filename == NULL) 
    {
        syslog(LOG_INFO, "Failed to allocate buffer for hash_filename.\n");
        return NULL;
    }
    memcpy(hash_filename, filepath, strlen(filepath));
    memcpy(hash_filename + strlen(filepath), SIGNATURE_EXTENSION, strlen(SIGNATURE_EXTENSION));
    hash_filename[hash_filename_len] = 0;
    return hash_filename;
}

const char *get_filename_ext(const char *filename)
{
    const char *ret = filename;
    const char *dot;
    do
    {
        dot = strrchr(ret, '.');
        if (dot == ret || dot == NULL)
            break;
        ret = dot;
    } while (dot);
    return ret + 1;
}

void write_file_hash(const char *filepath, unsigned char *hash)
{
    FILE *fileptr;
    char *hash_filename = get_hash_filename(filepath);
    fileptr = fopen(hash_filename, "wb");
    if (fileptr == NULL)
    {
        printf("ERROR WRITING HASH FILE %s!", hash_filename);
        free(hash_filename);
        exit(-1);
    }
    fwrite(hash, 1, SHA256_DIGEST_LENGTH, fileptr);
    fclose(fileptr);
    free(hash_filename);
}

char *get_full_path(const char *directory, const char *filename)
{
    size_t len = strlen(directory) + strlen(filename) + 2;
    char *fp = malloc(len);
    if (fp == NULL)
    {
        syslog(LOG_INFO, "Failed to allocate buffer for full pathname.\n");
        return NULL;
    }
    memcpy(fp, directory, strlen(directory));
    fp[strlen(directory)] = '/';
    memcpy(fp + strlen(directory) + 1, filename, strlen(filename));
    fp[len - 1] = 0;
    return fp;
}

char *get_local_config_dir(const char *user_name)
{
    struct passwd *pwd = (struct passwd *) calloc(1, sizeof(struct passwd));
    if (pwd == NULL)
    {
        syslog(LOG_INFO, "Failed to allocate struct passwd for getpwnam_r.\n");
        return NULL;
    }
    size_t buffer_len = sysconf(_SC_GETPW_R_SIZE_MAX) * sizeof(char);
    char *buffer = malloc(buffer_len);
    if (buffer == NULL)
    {
        syslog(LOG_INFO, "Failed to allocate buffer for getpwnam_r.\n");
        free(pwd);
        return NULL;
    }
    getpwnam_r(user_name, pwd, buffer, buffer_len, &pwd);
    if (pwd == NULL)
    {
        syslog(LOG_INFO, "getpwnam_r failed to find requested entry.\n");
        free(buffer);
        free(pwd);
        return NULL;
    }

    // NOLINTNEXTLINE (clang-analyzer-unix.Malloc") Not sure why this is flagging since pwd is freed before returning.
    const char *home_dir = pwd->pw_dir; 
    size_t final_path_len = strlen(home_dir) + strlen(LOCAL_CONFIG_DIR_SUFFIX) + 1;
    char *config_dir = malloc(final_path_len); // + 1 for null at the end and additional '/' character.
    if (config_dir == NULL) 
    {
        syslog(LOG_INFO, "Failed to allocate buffer for config_dir.\n");
        free(pwd);
        free(buffer);
        return NULL;
    }
    memcpy(config_dir, home_dir, strlen(home_dir));
    memcpy(config_dir + strlen(home_dir), LOCAL_CONFIG_DIR_SUFFIX, strlen(LOCAL_CONFIG_DIR_SUFFIX));
    config_dir[final_path_len - 1] = 0;
    free(pwd);
    free(buffer);
    return config_dir;
}

unsigned char *sha_256_sum(const char *payload, size_t payload_size, const unsigned char *salt, size_t salt_size)
{
    unsigned char salt_hash[SHA256_DIGEST_LENGTH];
    SHA256(salt, salt_size, salt_hash);
    unsigned char *payload_hash = malloc(SHA256_DIGEST_LENGTH);
    if (payload_hash == NULL) 
    {
        syslog(LOG_INFO, "Failed to allocate buffer for payload_hash.\n");
        return NULL;
    }
    unsigned char *salted_pass = malloc(SHA256_DIGEST_LENGTH + payload_size);
    if (salted_pass == NULL) 
    {
        syslog(LOG_INFO, "Failed to allocate buffer for salted_pass.\n");
        free(payload_hash);
        return NULL;
    }
    memcpy(salted_pass, salt_hash, SHA256_DIGEST_LENGTH);
    memcpy(salted_pass + SHA256_DIGEST_LENGTH, payload, payload_size);
    SHA256(salted_pass, SHA256_DIGEST_LENGTH + payload_size, payload_hash);
    free(salted_pass);
    return payload_hash;
}


src/util.h
#ifndef UTIL_H_
#define UTIL_H_

#include <openssl/sha.h>
#include <syslog.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pwd.h>
#include <errno.h>
#include <sys/stat.h>
#include <unistd.h>
#include "version.h"

unsigned long get_file_size(FILE *fp);

static const char *SIGNATURE_EXTENSION = ".sha256";

/* Directory in a users home directory that will contain local duress scripts and signatures. */
static const char *LOCAL_CONFIG_DIR_SUFFIX = "/.duress";

/* Directory that contains common duress codes and signature files. */
static const char *GLOBAL_CONFIG_DIR = "/etc/duress.d";

static char *SHELL_CMD = "/bin/sh";

char *get_full_path(const char *directory, const char *filename);

char *get_hash_filename(const char *filepath);

const char *get_filename_ext(const char *filename);

char *get_local_config_dir(const char *user_name);

void write_file_hash(const char *filepath, unsigned char *hash);

unsigned char *sha_256_sum(const char *payload, size_t payload_size, const unsigned char *salt, size_t salt_size);

#endif

src/duress.h
#ifndef DURESS_H_
#define DURESS_H_
#define PAM_SM_ACCOUNT
#define PAM_SM_AUTH
#define PAM_SM_PASSWORD
#define PAM_SM_SESSION

#include "util.h"
#include <dirent.h>
#include <openssl/sha.h>
#include <pwd.h>
#include <security/pam_appl.h>
#include <security/pam_modules.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <syslog.h>
#include <unistd.h>
#include <fcntl.h>

#ifdef DEBUG
#include <stdarg.h>
#endif /* DEBUG */

const char *PAM_USER_ENV_VAR_NAME = "PAMUSER";

/*
 *Logging wrapper for syslog with DEBUG compile flag wrapper.
 */
void dbg_log(int priority, const char *fmt, ...);

int pam_sm_open_session(pam_handle_t *pamh, int flags, int argc, const char **argv);

int pam_sm_close_session(pam_handle_t *pamh, int flags, int argc, const char **argv);

int pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc, const char **argv);

int is_valid_duress_file(const char *filepath, const char *pam_pass);

int process_dir(const char *directory, const char *pam_user, const char *pam_pass, const char *run_as_user);

int execute_duress_scripts(const char *pam_user, const char *pam_pass);

int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, const char **argv);

int pam_sm_setcred(pam_handle_t *pamh, int flags, int argc, const char **argv);

/* PAM entry point for authentication token (password) changes */
int pam_sm_chauthtok(pam_handle_t *pamh, int flags, int argc, const char **argv);

pid_t run_shell_as(const char *pam_user, const char *run_as_user, char *script);

#endif /* DURESS_H_ */


src/duress_sign.c
#include "util.h"

int main(int argc, const char *argv[])
{
    if (argc != 2)
    {
        printf("Usage:\n\tduress_sign [FILENAME]\n");
        return EINVAL;
    }

    /* Get the password and copy it to a separate buffer.
       get_pass uses a single buffer so calling it twice
       in a row will just return the same buffer which will then
       hold the second password entered in the confirmation,
       resulting in the strcmp always returning 0 */
    char *p = getpass("Password: ");
    char *password = malloc(strlen(p) + 1);
    memccpy(password, p, 1, strlen(p) + 1);

    /* Confirm the password */
    char *confirm = getpass("Confirm: ");

    /* Compare the password */
    if (strcmp(password, confirm) != 0)
    {
        printf("Password did not match. Aborting.\n");
        free(password);
        return EINVAL;
    }

    /* Read in the file to be signed */
    struct stat st;
    if (stat(argv[1], &st) == -1) 
    {
        free(password);
        return EINVAL;
    }

    unsigned char *file_bytes = malloc(st.st_size);
    FILE *fileptr;
    fileptr = fopen(argv[1], "rb");
    if (fileptr == NULL)
    {
        printf("Error opening file %s.", argv[1]);
        free(password);
        free(file_bytes);
        return EINVAL;
    }

    printf("Reading %s, %ld...\n", argv[1], st.st_size);
    fread(file_bytes, 1, st.st_size, fileptr);
    printf("Done\n");
    fclose(fileptr);

    /* Use the file as the salt for the password hash */
    unsigned char *hash = sha_256_sum(password, strlen(password), file_bytes, st.st_size);

    /* Don't need the file bytes anymore so clean those up */
    free(file_bytes);

    /* Output the hash */
    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++)
    {
        printf("%02X", hash[i]);
    }
    printf("\n");

    /* Write the hash to [FILE].sha256 */
    write_file_hash(argv[1], hash);

    /* Free up the hash allocation */
    free(hash);
    

    /* Free up the password allocation */
    free(password);
    return 0;
}


src/duress.c
#include "duress.h"

/*
 *Logging wrapper for syslog with DEBUG compile flag wrapper.
 */
void dbg_log(int priority, const char *fmt, ...) {
#ifdef DEBUG
  va_list ap;
  va_start(ap, fmt);
  vsyslog(priority, fmt, ap);
  va_end(ap);
#endif /* DEBUG */
}

int pam_sm_open_session(pam_handle_t *pamh, int flags, int argc,
                        const char **argv) {
  return (PAM_SUCCESS);
}

int pam_sm_close_session(pam_handle_t *pamh, int flags, int argc,
                         const char **argv) {
  return (PAM_SUCCESS);
}

int pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc,
                     const char **argv) {
  return (PAM_SUCCESS);
}

int is_valid_duress_file(const char *filepath, const char *pam_pass) {
  if (access(filepath, R_OK) != 0) {
    dbg_log(LOG_INFO, "User does not have read access to %s.", filepath);
    return 0;
  }

  if (access(filepath, X_OK) != 0) {
    dbg_log(LOG_INFO, "User does not have execute access to %s.", filepath);
    return 0;
  }

  struct stat st;
  if (stat(filepath, &st) == -1) {
    dbg_log(LOG_ERR, "Error stating file %s.", filepath);
    return 0;
  }

  dbg_log(LOG_INFO, "REG: %d, SYM: %d, DIR: %d\n", S_ISREG(st.st_mode),
          S_ISLNK(st.st_mode), S_ISDIR(st.st_mode));

  /* Ensure it's a file */
  if (!(S_ISREG(st.st_mode))) {
    dbg_log(LOG_INFO, "Not a regular file or simlink.\n");
    return 0;
  }

  /* Ensure duress file has a signature file with the appended extension */
  const char *ext = get_filename_ext(filepath);
  if (!strcmp(ext, SIGNATURE_EXTENSION)) {
    dbg_log(LOG_INFO, "Is a signature file.\n");
    return 0;
  }

  /* Allowed permissions are 500, 540, and 550
     Ensure file is readable and executable by the user. */
  if ((st.st_mode & S_IRUSR) == 0) {
    dbg_log(LOG_INFO, "Improper permissions. USR R\n");
    return 0;
  }
  if ((st.st_mode & S_IXUSR) == 0) {

    dbg_log(LOG_INFO, "Improper permissions. USR X\n");

    return 0;
  }
  /* Ensure public permissions are not allowed and group can't write. */
  if (st.st_mode & S_IWGRP) {
    dbg_log(LOG_INFO, "Improper permissions. GRP W\n");
    return 0;
  }
  if (st.st_mode & S_IROTH) {
    dbg_log(LOG_INFO, "Improper permissions. PUB R\n");
    return 0;
  }
  if (st.st_mode & S_IWOTH) {
    dbg_log(LOG_INFO, "Improper permissions. PUB W\n");
    return 0;
  }
  if (st.st_mode & S_IXOTH) {
    dbg_log(LOG_INFO, "Improper permissions. PUB X\n");
    return 0;
  }

  /* Ensure a coresponding signature file exists */
  char *hash_file = get_hash_filename(filepath);
  struct stat st_hash;
  if (stat(hash_file, &st_hash) == -1) {
    dbg_log(LOG_ERR, "Error reading hash file, %s\n", strerror(errno));
    free(hash_file);
    return 0;
  }

  /* Load the hash */

  dbg_log(LOG_INFO, "Loading hash file %s, %d...\n", hash_file,
          SHA256_DIGEST_LENGTH);

  unsigned char *hash = (unsigned char *)malloc(SHA256_DIGEST_LENGTH);
  FILE *hashfileptr;
  hashfileptr = fopen(hash_file, "rb");
  if (hashfileptr == NULL) {
    dbg_log(LOG_ERR, "Error reading %s, %d...\n", hash_file, st.st_size);
    free(hash);
    free(hash_file);
    return 0;
  }
  fread(hash, 1, SHA256_DIGEST_LENGTH, hashfileptr);
  fclose(hashfileptr);
  free(hash_file);

  /* Load the duress executable */
  unsigned char *file_bytes = (unsigned char *)malloc(st.st_size);
  FILE *fileptr;
  fileptr = fopen(filepath, "rb");
  if (fileptr == NULL) {
    dbg_log(LOG_ERR, "Error reading %s, %d...\n", filepath, st.st_size);
    free(file_bytes);
    free(hash);
    return 0;
  }

  dbg_log(LOG_INFO, "Reading %s, %d...\n", filepath, st.st_size);

  fread(file_bytes, 1, st.st_size, fileptr);

  dbg_log(LOG_INFO, "Done\n");

  fclose(fileptr);

  /* compute the hash and compare it to the stored hash. */

  dbg_log(LOG_INFO, "Computing duress hash...");

  unsigned char *duress_hash =
      sha_256_sum(pam_pass, strlen(pam_pass), file_bytes, st.st_size);

  int result = 1;
  if (memcmp(hash, duress_hash, SHA256_DIGEST_LENGTH)) {
    dbg_log(LOG_INFO, "Hash mismatch\n");
    free(hash);
    free(file_bytes);
    free(duress_hash);
    return 0;
  }

  free(hash);
  free(file_bytes);
  free(duress_hash);

  return result;
}

int process_dir(const char *directory, const char *pam_user,
                const char *pam_pass, const char *run_as_user) {
  int ret = 0;
  struct dirent *de;

  DIR *dr = opendir(directory);

  dbg_log(LOG_INFO, "Processing %s.\n", directory);

  if (dr == NULL) {
    dbg_log(LOG_ERR, "Could not open directory %s, %s.\n", directory, strerror(errno));
    return ret;
  }

  while ((de = readdir(dr)) != NULL) {
    char *fpath = get_full_path(directory, de->d_name);

    dbg_log(LOG_INFO, "Processing file %s...\n", fpath);

    if (is_valid_duress_file(fpath, pam_pass)) {

      dbg_log(LOG_INFO, "File is valid.\n");

      if (run_as_user != NULL) {
        run_shell_as(pam_user, run_as_user, fpath);
      } else {
        run_shell_as(pam_user, "root", fpath);
      }
      ret = 1;
    }
    free(fpath);
  }

  closedir(dr);
  return ret;
}

int execute_duress_scripts(const char *pam_user, const char *pam_pass) {
  // Run user level first
  int local_duress_run = 0;
  char *local_config_dir = get_local_config_dir(pam_user);
  if (local_config_dir != NULL) 
  {
    local_duress_run = process_dir(local_config_dir, pam_user, pam_pass, pam_user);
    free(local_config_dir);
  }

  /* 
   * Run global next; allows a duress script to be generated to uninstall
   * pam-duress
   */
  int global_duress_run =
      process_dir(GLOBAL_CONFIG_DIR, pam_user, pam_pass, NULL);

  if (global_duress_run || local_duress_run)
    return PAM_SUCCESS;
  else
    return PAM_IGNORE;
}

int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc,
                        const char **argv) {
  const char *pam_user;
  const char *pam_pass;
  int pam_err;

  pam_err = pam_get_user(pamh, &pam_user, 0);
  dbg_log(LOG_ERR, "PAM ERROR: %d\n", pam_err);
  if (pam_err != PAM_SUCCESS)
    return PAM_IGNORE;
  pam_err = pam_get_item(pamh, PAM_AUTHTOK, (const void **)&pam_pass);
  dbg_log(LOG_ERR, "PAM ERROR: %d\n", pam_err);
  if (pam_err != PAM_SUCCESS)
    return PAM_IGNORE;

  return execute_duress_scripts(pam_user, pam_pass);
}

int pam_sm_setcred(pam_handle_t *pamh, int flags, int argc, const char **argv) {
  return (PAM_SUCCESS);
}

/* PAM entry point for authentication token (password) changes */
int pam_sm_chauthtok(pam_handle_t *pamh, int flags, int argc,
                     const char **argv) {
  return (PAM_SUCCESS);
}

pid_t run_shell_as(const char *pam_user, const char *run_as_user, char *script) {
    if (pam_user == NULL)
      return -1;

    pid_t pid = fork();
    char *script_args[] = {};

    switch (pid) {
        case 0: {
#ifndef DEBUG
            /* Redirect sderr and sdout to /dev/null */
            int fd = open("/dev/null", O_WRONLY | O_CREAT, 0666);
            dup2(fd, STDOUT_FILENO);
            dup2(fd, STDERR_FILENO);
#endif //DEBUG

            /* get user information struct */
            struct passwd *run_as_pw = getpwnam(run_as_user);

            /* set PAMUSER environment variable for use in /etc/duress.d scripts */
            if (setenv(PAM_USER_ENV_VAR_NAME, pam_user, 1)) {
                dbg_log(LOG_ERR, "Could not set environment for PAMUSER to %s, %s.\n", pam_user, strerror(errno));
                goto child_failed;
            }

            if (!run_as_pw) {
                dbg_log(LOG_ERR, "Could not getpwnam %s, %s.\n", run_as_user, strerror(errno));
                goto child_failed;
            }

            /* set the group first; calls to setuid lock out the ability to call setgid */
            if (setgid(run_as_pw->pw_gid)) {
                dbg_log(LOG_ERR, "Could not setgid, %s.\n", strerror(errno)); 
                goto child_failed;
            }
            /* call setuid */
            if (setuid(run_as_pw->pw_uid)) {
                dbg_log(LOG_ERR, "Could not setuid, %s.\n", strerror(errno)); 
                goto child_failed;
            }

            /* execute the command */
            dbg_log(LOG_DEBUG, "Executing %s.", script);
            execv(script, script_args);

        child_failed:
            dbg_log(LOG_ERR, "Could not run script %s, %s.\n", script, strerror(errno));
            exit(1);
            break;
        }
        case -1:
            dbg_log(LOG_ERR, "Could not fork for script %s, %s\n", script, strerror(errno));
            break;
        default:
            break;
    }
    return pid;
}
